### PR TITLE
MVCLI Testcase for OpenPower Marvell adapters.

### PR DESCRIPTION
**Pull request by pridhiviraj**
Mvcli Test for OpenPower based Marvell adapter 88SE9230.

This tests the different mvcli tool operations like
info, get, set, smart, locate, register, delete, create, event
on RAID Marvell adapter 88SE9230

Signed-off-by: Pridhiviraj Paidipeddi <ppaidipe@linux.vnet.ibm.com>

### CODE REVIEW

File: io/disk/mvcli_test.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+        :param pd_ids: List of phyisical disk ID's under adapter id adapter_id
+        """
+        if 'ppc64le' not in platform.processor():
+            self.cancel("Processor is not ppc64le")
+        self.tool_url = self.params.get('tool_url', '*', default=None)
+        self.adapter_id = self.params.get('adapter_id', '*', default=None)
+        self.pd_ids = self.params.get('pd_ids', '*', default=None).split(",")
+        path = self.fetch_asset("mvcli", locations=[self.tool_url],
+                                expire="7d")
+        self.run_command("cp %s %s" % (path, self.teststmpdir))
```
Comment (narasimhan-v) [inline io/disk/mvcli_test.py]: please use `shutil` library

File: io/disk/mvcli_test.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+    def test_set(self):
+        """
+        Set configuration parameters of VD, PD, HBA or Driver.
+        """
+        self.log.info("Mvcli set configuration parameters test")
+        self.run_command("%s set -o hba -i %s --rawupdateon" % (self.base,
+                         self.adapter_id))
+        get_cmd = "%s get -o hba -i %s" % (self.base, self.adapter_id)
+        output = self.run_cmd_output(get_cmd)
+        if "Raw Update:              on" in output:
```
Comment (narasimhan-v) [inline io/disk/mvcli_test.py]: This is not the best parsing. My suggestion:

`if "Raw Update:" in output and "on" in output:`

File: io/disk/mvcli_test.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+        self.run_command("%s get -o hba" % self.base)
+        for pd in self.pd_ids:
+            self.run_command("%s get -o pd -i %s" % (self.base, pd))
+
+    def test_set(self):
+        """
+        Set configuration parameters of VD, PD, HBA or Driver.
+        """
+        self.log.info("Mvcli set configuration parameters test")
+        self.run_command("%s set -o hba -i %s --rawupdateon" % (self.base,
```
Comment (narasimhan-v) [inline io/disk/mvcli_test.py]:
```
    for state in ['on', 'off']:
        self.run_command("%s set -o hba -i %s --rawupdate%s" % (self.base,
                         self.adapter_id, state))
        get_cmd = "%s get -o hba -i %s" % (self.base, self.adapter_id)
        output = self.run_cmd_output(get_cmd)
        if "Raw Update:              %s" % state in output:
            self.log.info("Mvcli set --rawupdate%s operation passed \
                          on adapter %s", state, self.adapter_id)
        else:
            self.fail("Mvcli set --rawupdate%s operation failed \
                          on adapter %s", state, self.adapter_id)
```

File: io/disk/mvcli_test.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+        Delete virtual disk or spare drive.
+        """
+        self.log.info("Mvcli delete VD test")
+        output = self.run_cmd_output("%s info -o vd" % self.base)
+        if "No virtual disk is found." in output:
+            self.log.info("No VD's are created yet/ skipping delete \
+                          VD operation")
+            return
+        vd_list = []
+        search_string = r"id:                  (\d)"
```
Comment (narasimhan-v) [inline io/disk/mvcli_test.py]: please use better regex

File: io/disk/mvcli_test.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+            elif level == "1":
+                if pd_list_len < 2:
+                    continue
+                pd_list = "%s,%s" % (self.pd_ids[0], self.pd_ids[1])
+            elif level == "10":
+                if pd_list_len < 4:
+                    continue
+                pd_list = "%s,%s,%s,%s" % (self.pd_ids[0], self.pd_ids[1],
+                                           self.pd_ids[2], self.pd_ids[3])
+            else:
```
Comment (narasimhan-v) [inline io/disk/mvcli_test.py]: We are only forming the list of raid levels. So, I do not think this is needed

File: io/disk/mvcli_test.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+        Get the current events.
+        """
+        self.log.info("Mvcli event test")
+        self.run_command("%s event -s 10" % self.base)
+
+    def tearDown(self):
+        """
+        Gather any kernels errors at the end of test
+        """
+        self.log.info("Gathering any kernel errors")
```
Comment (narasimhan-v) [inline io/disk/mvcli_test.py]: Removing any created vds or RAIDs can be taken care in `tearDown()`

File: io/disk/mvcli_test.py (hunk 2 above)
Comment (pridhiviraj) [inline io/disk/mvcli_test.py]: Here output is a multi line string not single line string.

File: io/disk/mvcli_test.py (hunk 5 above)
Comment (pridhiviraj) [inline io/disk/mvcli_test.py]: Based on the RAID level, list of pd's needed will vary. So list of pd's to be selected is needed based on RAID level

File: io/disk/mvcli_test.py (hunk 6 above)
Comment (pridhiviraj) [inline io/disk/mvcli_test.py]: possibly test_delete we can call again in tearDown.

**Comment by pridhiviraj** [comment]
@narasimhan-v  Thanks for the review, Please look at the latest one with fixes for your review. And also added FW flash test as well.

File: io/disk/mvcli_fw_flash.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+import shutil
+import platform
+from avocado import Test
+from avocado import main
+from avocado.utils import process
+from avocado.utils.process import CmdError
+
+
+class MvcliFwFlash(Test):
```
Comment (narasimhan-v) [inline io/disk/mvcli_fw_flash.py]: Please have a look at nvmetest.py on how firmware flash is handled in the same test as other tests

**Comment by pridhiviraj** [comment]
@narasimhan-v as per suggestion merged flash test with main testcase with optional run. And also fixed remaining comments. Please check now. Thanks.

File: io/disk/mvcli_test.py [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+        self.run_command("dmesg -T --level=alert,crit,err,warn")
+        self.test_delete()
```
Comment (narasimhan-v) [inline io/disk/mvcli_test.py]: @pridhiviraj small change. `test_delete()` should be called before collecting dmesg logs

Comment (pridhiviraj) [inline io/disk/mvcli_test.py]: okay, will do it. Thanks

File: io/disk/mvcli_test.py [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+
+    def test_register(self):
+        """
+        operation the register on adapter.
+        test read/write mailbox registers on a test adapter
+        """
+        self.log.info("Mvcli register read/write test")
+        registers = ["0", "1"]
+        values = ["1", "0"]
+        for reg in registers:
```
Comment (PraveenPenguin) [inline io/disk/mvcli_test.py]: here as well 
`
 for reg in ["0", "1"]:`

File: io/disk/mvcli_test.py [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+            for value in values:
```
Comment (PraveenPenguin) [inline io/disk/mvcli_test.py]: here also please 

` for value in ["1", "0"]:`

File: io/disk/mvcli_test.py [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+            obj = re.search(search_string, line)
+            if obj:
+                vd_list.append(obj.group(1))
+        if vd_list is None:
+            self.log.info("Not able to found any VD's are found")
+        self.log.info("List of VD's to be deleted: %s", vd_list)
+        for vd in vd_list:
+            self.run_command("%s delete -o vd -i %s" % (self.base_force, vd))
+        output = self.run_cmd_output("%s info -o vd" % self.base)
+        if "No virtual disk is found." in output:
```
Comment (PraveenPenguin) [inline io/disk/mvcli_test.py]: here also please

File: io/disk/mvcli_test.py [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+        and try to delete VD which was created
+        for RAID 0, Use all disks from PD,
+        for RAID 1, Use minimum of two disks from PD
+        for RAID 10, Use minimum of four disks from PD
+        """
+        raid_levels = ["0", "1", "10"]
+        base = "%s create -o vd " % self.base_force
+        pd_list_len = len(self.pd_ids)
+        self.log.info("PD List %s, length: %s", self.pd_ids, pd_list_len)
+        for level in raid_levels:
```
Comment (PraveenPenguin) [inline io/disk/mvcli_test.py]: here also please 
` for level in ["0", "1", "10"]`

**Comment by PraveenPenguin** [comment]
@pridhiviraj please attach job log as well as i don't have setup

**Comment by pridhiviraj** [comment]
avocado run mvcli_test.py -m mvcli_test.py.data/mvcli.yaml
JOB ID     : da81404ec2cadaa46dfa4e4d37b84a09c6e5c991
JOB LOG    : /root/avocado/job-results/job-2017-07-13T08.25-da81404/job.log
 (01/11) mvcli_test.py:MvcliTest.test_help;run-f0fb: PASS (0.08 s)
 (02/11) mvcli_test.py:MvcliTest.test_flash;run-f0fb: PASS (13.84 s)
 (03/11) mvcli_test.py:MvcliTest.test_info;run-f0fb: PASS (0.30 s)
 (04/11) mvcli_test.py:MvcliTest.test_get;run-f0fb: PASS (0.17 s)
 (05/11) mvcli_test.py:MvcliTest.test_set;run-f0fb: PASS (0.37 s)
 (06/11) mvcli_test.py:MvcliTest.test_smart;run-f0fb: PASS (2.22 s)
 (07/11) mvcli_test.py:MvcliTest.test_locate;run-f0fb: PASS (0.38 s)
 (08/11) mvcli_test.py:MvcliTest.test_register;run-f0fb: PASS (0.31 s)
 (09/11) mvcli_test.py:MvcliTest.test_delete;run-f0fb: PASS (0.07 s)
 (10/11) mvcli_test.py:MvcliTest.test_create;run-f0fb: PASS (14.63 s)
 (11/11) mvcli_test.py:MvcliTest.test_event;run-f0fb: PASS (0.84 s)
RESULTS    : PASS 11 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 33.90 s

**Comment by pridhiviraj** [comment]
[job.txt](https://github.com/avocado-framework-tests/avocado-misc-tests/files/1145101/job.txt)

**Comment by PraveenPenguin** [comment]
@narasimhan-v thanks for review

**Pull request merged**